sycl/unittests/device/aspect-selector.cpp

```diff
@@ -0,0 +1,187 @@
+
+#include <vector>
+
+using namespace sycl;
+
+std::vector<aspect> Required, Denied;
+std::vector<std::vector<aspect>> Available;
+
+pi_result redefinedDeviceRetain(pi_device) { return PI_SUCCESS; }
+pi_result redefinedDeviceRelease(pi_device) { return PI_SUCCESS; }
+
+platform *Platform;
+
+pi_result redefinedDeviceGetInfo(pi_device device, pi_device_info param_name,
+                                 size_t param_value_size, void *param_value,
+                                 size_t *param_value_size_ret) {
+  long DevIdx = reinterpret_cast<long>(device) - 1;
+
+  assert(DevIdx >= 0 && (unsigned)(DevIdx) < Available.size());
+
+  const std::vector<aspect> &Aspects = Available[DevIdx];
+
+  if (param_name == PI_DEVICE_INFO_TYPE) {
+    static const std::map<aspect, pi_device_type> Types = {
+        {aspect::cpu, PI_DEVICE_TYPE_CPU},
+        {aspect::gpu, PI_DEVICE_TYPE_GPU},
+        {aspect::accelerator, PI_DEVICE_TYPE_ACC}};
+
+    if (param_value) {
+      bool DeviceTypeSet = false;
+      for (const auto &P : Types) {
+        auto It = std::find(Aspects.begin(), Aspects.end(), P.first);
+        if (It != Aspects.end()) {
+          *reinterpret_cast<pi_device_type *>(param_value) = P.second;
+          DeviceTypeSet = true;
+          break;
+        }
+      }
+
+      assert(DeviceTypeSet);
+    }
+
+    return PI_SUCCESS;
+  }
+
+  if (param_name == PI_DEVICE_INFO_PARENT_DEVICE) {
+    if (param_value)
+      *reinterpret_cast<pi_device *>(param_value) = NULL;
+
+    return PI_SUCCESS;
+  }
+
+  if (param_name == PI_DEVICE_INFO_EXTENSIONS) {
+    if (param_value_size_ret)
+      *param_value_size_ret = 0;
+    return PI_SUCCESS;
+  }
+
+  if (param_name == PI_DEVICE_INFO_PLATFORM) {
+    if (param_value)
+      *reinterpret_cast<pi_platform *>(param_value) =
+          detail::getSyclObjImpl(*Platform)->getHandleRef();
+    return PI_SUCCESS;
+  }
+
+  if (param_name == PI_DEVICE_INFO_NAME) {
+    const std::string DevName = "Test Device # " + std::to_string(DevIdx);
+    if (param_value_size_ret)
+      *param_value_size_ret = DevName.length();
+    if (param_value) {
+      size_t L = std::min(DevName.length(), param_value_size);
+      memcpy(param_value, DevName.data(), L);
+    }
+
+    return PI_SUCCESS;
+  }
+
+  return PI_INVALID_VALUE;
+}
+
+pi_result redefinedDevicesGet(pi_platform Plt, pi_device_type DeviceType,
+                              pi_uint32 NumEntries, pi_device *Devs,
+                              pi_uint32 *NumDevices) {
+  if (NumDevices)
+    *NumDevices = Available.size();
+
+  if (Devs) {
+    long MaxIdx = std::min((size_t)NumEntries, Available.size());
+
+    for (int DevIdx = 0; DevIdx < MaxIdx; ++DevIdx)
+      Devs[DevIdx] = reinterpret_cast<pi_device>(DevIdx + 1);
+  }
+
+  return PI_SUCCESS;
+}
+
+class AspectSelector : public ::testing::Test {
+public:
+  AspectSelector() = default;
+
+protected:
+  static void SetUpTestSuite() {
+    MPlt = new platform{sycl::default_selector()};
```

Comment:
I would prefer using smart pointers here despite of explicit object destruction in `TearDownTestSuite` function, but I guess it's not critical.